ambry-store/src/test/java/com.github.ambry.store/DiskSpaceAllocatorTest.java

```diff
@@ -0,0 +1,397 @@
+   * "conc-test-{n}"
+   * @param requirementsList The list of {@link DiskSpaceRequirements} used to determine the number of operations to
+   *                         start
+   * @param allocate {@code true} if each task should allocate a new segment.
+   * @param free {@code true} if each task should free a (hopefully) existing segment.
+   */
+  private void runConcurrencyTest(List<DiskSpaceRequirements> requirementsList, final boolean allocate,
+      final boolean free) throws Exception {
+    List<Callable<Void>> tasks = new ArrayList<>();
+    final String filenamePrefix = "conc-test-";
+    int fileCount = 0;
+    for (final DiskSpaceRequirements requirements : requirementsList) {
+      for (int i = 0; i < requirements.getSegmentsNeeded(); i++, fileCount++) {
+        final String filename = filenamePrefix + fileCount;
+        tasks.add(() -> {
+          if (allocate) {
+            allocateAndVerify(filename, requirements.getSegmentSizeInBytes());
+          }
+          if (free) {
+            freeAndVerify(new File(allocatedFileDir, filename), requirements.getSegmentSizeInBytes());
+          }
+          return null;
+        });
+        fileCount++;
+      }
+    }
+    for (Future<Void> future : exec.invokeAll(tasks)) {
+      future.get(10, TimeUnit.SECONDS);
+    }
+  }
+
+  /**
+   * Verify that the layout of the pool matches the expected state.
+   * @param expectedState a map from file size to the number of reserve files expected of that size.
+   */
+  private void verifyPoolState(Map<Long, Integer> expectedState) {
+    assertEquals("Wrong number of file size dirs", expectedState.size(), reserveFileDir.list().length);
+    for (Map.Entry<Long, Integer> entry : expectedState.entrySet()) {
+      long size = entry.getKey();
+      int count = entry.getValue();
+      File fileSizeDir = new File(reserveFileDir, Long.toString(size));
+      String[] filenameList = fileSizeDir.list();
+      assertNotNull("Error while listing files for size " + size, filenameList);
+      assertEquals("Wrong number of files for size " + size, count, filenameList.length);
+    }
+  }
+
+  private DiskSpaceAllocator constructAllocator() throws StoreException {
+    return new DiskSpaceAllocator(reserveFileDir, requiredSwapSegmentsPerSize);
+  }
+
+  /**
+   * A builder for a map that describes the number of reserve files for each file size.
+   * Used with {@link #verifyPoolState(Map)}
+   */
+  private static class ExpectedStateBuilder {
+    private final Map<Long, Integer> map = new HashMap<>();
+
+    /**
+     * Update the map with an entry for a certain file size.
+     * @param size the file size.
+     * @param count the expected number of files
+     * @return this builder
+     */
+    ExpectedStateBuilder add(long size, int count) {
+      map.put(size, count);
+      return this;
+    }
+
+    /**
+     * @return the {@link Map}
+     */
+    Map<Long, Integer> build() {
```

Comment:
you can avoid this function and just make the variable non-private (internal class anyway)